src/main/java/com/google/cloud/tools/ide/login/GoogleLoginState.java

```diff
@@ -428,104 +400,79 @@ public void simulateLoginStatusChange(boolean login) {
     uiFacade.notifyStatusIndicator();
   }
 
-  private void updateUserCredentials(GoogleTokenResponse tokenResponse) {
+  private void logInHelper(GoogleTokenResponse tokenResponse) {
     refreshToken = tokenResponse.getRefreshToken();
     accessToken = tokenResponse.getAccessToken();
     oAuth2Credential = makeCredential();
-    accessTokenExpiryTime = System.currentTimeMillis() / 1000
-        + tokenResponse.getExpiresInSeconds().longValue();
+    accessTokenExpiryTime = System.currentTimeMillis() / 1000 + tokenResponse.getExpiresInSeconds();
     email = queryEmail();
-    saveCredentials();
+    persistCredentials();
     uiFacade.notifyStatusIndicator();
     notifyLoginStatusChange(true);
   }
 
   private void retrieveSavedCredentials() {
+    Preconditions.checkState(!isLoggedIn(), "Should be called only once in the constructor.");
 
     OAuthData savedAuthState = authDataStore.loadOAuthData();
 
-    // the stored email can be null in the case where the external browser
-    // was launched, because we can't extract the email from the external
-    // browser
     if (savedAuthState.getRefreshToken() == null || savedAuthState.getStoredScopes().isEmpty()) {
       authDataStore.clearStoredOAuthData();
       return;
     }
 
-    accessToken = savedAuthState.getAccessToken();
-    refreshToken = savedAuthState.getRefreshToken();
-    accessTokenExpiryTime = savedAuthState.getAccessTokenExpiryTime();
-    this.email = savedAuthState.getStoredEmail();
-
-    isLoggedIn = true;
-
     if (!oAuthScopes.equals(savedAuthState.getStoredScopes())) {
```

Comment:
Technically this is only required if `oAuthScopesScopes` ⊈ `savedScopes`, not just full equality.  But this case is pretty unlikely.